src/keyboard.rs

```diff
@@ -0,0 +1,1445 @@
+    ///
+    /// Each flag represents a modifier and is set if this modifier is active.
+    #[derive(Default)]
+    pub struct ModifiersState: u32 {
+        // left and right modifiers are currently commented out, but we should be able to support
+        // them in a future release
+        /// The "shift" key.
+        const SHIFT = 0b100;
+        // const LSHIFT = 0b010 << 0;
+        // const RSHIFT = 0b001 << 0;
+        /// The "control" key.
+        const CONTROL = 0b100 << 3;
+        // const LCTRL = 0b010 << 3;
+        // const RCTRL = 0b001 << 3;
+        /// The "alt" key.
+        const ALT = 0b100 << 6;
+        // const LALT = 0b010 << 6;
+        // const RALT = 0b001 << 6;
+        /// This is the "windows" key on PC and "command" key on Mac.
+        const SUPER = 0b100 << 9;
+        // const LSUPER  = 0b010 << 9;
+        // const RSUPER  = 0b001 << 9;
+    }
+}
+
+#[cfg(feature = "serde")]
+mod modifiers_serde {
+    use super::ModifiersState;
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+    #[derive(Default, Serialize, Deserialize)]
+    #[serde(default)]
+    #[serde(rename = "ModifiersState")]
+    pub struct ModifiersStateSerialize {
+        pub shift_key: bool,
+        pub control_key: bool,
+        pub alt_key: bool,
+        pub super_key: bool,
+    }
+
+    impl Serialize for ModifiersState {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            let s = ModifiersStateSerialize {
+                shift_key: self.shift_key(),
+                control_key: self.control_key(),
+                alt_key: self.alt_key(),
+                super_key: self.super_key(),
+            };
+            s.serialize(serializer)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for ModifiersState {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            let ModifiersStateSerialize {
+                shift_key,
+                control_key,
+                alt_key,
+                super_key,
+            } = ModifiersStateSerialize::deserialize(deserializer)?;
+            let mut m = ModifiersState::empty();
+            m.set(ModifiersState::SHIFT, shift_key);
+            m.set(ModifiersState::CONTROL, control_key);
+            m.set(ModifiersState::ALT, alt_key);
+            m.set(ModifiersState::SUPER, super_key);
+            Ok(m)
+        }
+    }
+}
+
+/// Contains the platform-native physical key identifier (aka scancode)
+#[derive(Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum NativeKeyCode {
+    Unidentified,
+    Windows(u16),
+    MacOS(u16),
+    XkbCode(u32),
+    XkbSym(u32),
+    Web(String),
+
+    /// This is the android "key code" of the event as returned by
+    /// `KeyEvent.getKeyCode()`
+    Android(u32),
+}
```

Comment:
What is the purpose of this?

- To allow matching key-press/release events: does not require that this type is public
- To allow serialisation: also does not require that this is public
- To allow users to encode/decode non-standard keys: is this sufficiently useful?

I mean, this is *functional*, but locks winit into a complex public API where something much simper would suffice most use-cases (e.g. make `NativeKeyCode` a struct with private details).

Reply:
> (e.g. make `NativeKeyCode` a struct with private details).

Sure. I'm fine with going with this approach.

Just to clarify, would the following be okay with you?

```rust
#[derive(Clone, PartialEq, Eq, Hash)]
#[derive(Serialize, Deserialize)]
enum NativeKeyCodeData {
    Unidentified,
    Windows(u16),
    MacOS(u16),
    ...
}

#[derive(Clone, PartialEq, Eq, Hash)]
#[derive(Serialize, Deserialize)]
pub struct NativeKeyCode(NativeKeyCodeData); // NativeKeyCode is `pub` but NativeKeyCodeData isn't
```

Reply:
On further thought, this will just make the implementation more obscure, and we don't gain anything by not making that information `pub`, since if this is serialized than any change to the implementation will break existing code anyways.

So it's better to make it public and it's at least easier to detect that we made a breaking change.

(Everything that can be serialized is also "public")

Reply:
This isn't quite true, even when simply deriving serde: e.g. the string representation can still be changed without breaking serialisation.

Perhaps the strangest thing is the use of `Key::Character(&'static str)` (with leaked strings) in other parts of this API with `String` here. Still, the string reprs can be tweaked in a new winit release anyway (a breaking change, but not a big issue).

Reply:
It's mostly for enabling keybinds to "just work" if users have keyboards which emit keycodes we don't handle. `Unidentified` is specifically for cases where we literally cannot identify the key being pressed. I don't recall the exact edge-case justifying its inclusion, but it's either that Windows likes to sometimes emit 0 as the keycode, or that the Web API folds all unidentified cases into `key = "Unidentified"`, or both. It's been a while, heh.

I'm also reminded now that I wanted to split out `XkbSym`, as it's kind of the odd one out in this enum (being a keysym, and not a keycode). We could perhaps make it simpler by only having 3 enum variants, `Native`, `Web`, and `Unidentified`.

Reply:
We want to allow applications to bind non-standard keys to actions. If such bindings are saved and synchronized across platforms, then we need to be careful, because we can't assume that a non-standard key will generate the same identifier on all platforms. That's why we also identify the platform along with the key-code.